README.md
# ultraInvoice
Re-create Invoice and monthly bill from AWS MR file


docs/ReleaseNotes.txt
1.0
1. [S架构]输入数据改用AWS月度报告文件，多代付账户的AWS代付月度报告汇总后统一处理。
2. [R需求]可针对特定客户的特定账号设置泰岳折扣率，出账时生成折扣费用

2.0
1. [S]改为AWS Lambda启动、每月定期运行。调度更灵活，成本更低。
2. [S]日志输出到AWS Cloud Watch Logs。更便于监控运行故障。

2.1
1. [R]收据从随机数编号改为顺序编号
2. [Bug]中手游xls收据和账单数据不对[原始月度账单数据问题，194461301570-aws-billing-csv-2019-09.csv的内容也包含中手游账户了]
3. [R]输入文件改用带标签的月度报告
4. [R]月账单增加Region层，尽量贴近系统中层次分类

2.2
- [Request]汇率取小数点后4位。收据中最后一行显示税率：17%
- [R]收据编号中日期从账单月当改为下月
- [R]去掉Invoice页眉页脚，方便打印
- [R]按照总账单百分百收取Partner-led Support费用，优先使用customer的SupportRate属性，其次使用Account的
- [Bug]使收据中"其他"在所有正值行之后
- [B]解决汇总详单RI/OD混合问题[只按产品和usage汇总不对]
- [Optimize] 优化详单层次，去掉意义不大的第4层，同时尽量靠近AWS系统中账单；
- [O]泰岳折扣、Credit等负值的行对应No Region；
- [O]泰岳折扣优先使用customer的Discount属性，其次使用Account的
- [O]详单Sheet设置首行冻结

TBD
- [R]按AWS Support费用打折收取Partner-led Support费用，优先使用customer的SupportDiscount属性，其次使用Account的
- [O]excel 增加筛选

Mail.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# @File  : Mail.py
# @Author: Sui Huafeng
# @Date  : 2019/10/17
# @Desc  : 
#

import os
import smtplib
from email import encoders
from email.header import Header
from email.mime.base import MIMEBase
from email.mime.multipart import MIMEMultipart
from email.mime.text import MIMEText
from email.utils import formataddr
import config as cfg


class Mail(object):
    Sender = 'yyw1@ultrapower.com.cn'  # 发件人邮箱账号
    Passwd = 'ultra#0917'
    MailServer = 'mail.ultrapower.com.cn'
    Port = '587'
    Bcc = ['suihf@ultrapower.com.cn']

    def __init__(self, subject='', text=''):
        self.sub = subject
        self.text = text
        self.server = smtplib.SMTP(self.MailServer, self.Port)  # 发件人邮箱中的SMTP服务器，端口是25
        self.server.login(self.Sender.split('@')[0], self.Passwd)  # 发件人邮箱账号、邮箱密码

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.server.quit()

    def send(self, receive, attachments, sub='', text='', ):
        try:
            msg = MIMEMultipart('related')
            msg['From'] = formataddr(["泰岳云业务", self.Sender])  # 发件人邮箱昵称、发件人邮箱账号
            msg['To'] = receive[0]  # 收件人邮箱昵称、收件人邮箱账号
            if len(receive) > 1:
                msg['Cc'] = ','.join(receive[1:])  # 收件人邮箱昵称、收件人邮箱账号
            msg['Subject'] = sub
            text += """<br><br><br><small>
            (如您收到的附件后缀为.dat, 是您的邮件客户端配置引起的，请保存并修改为.xls)</small><br><br>
            云业务事业部<br>
            神州泰岳软件股份有限公司<br>
            电话：+86(10) 58847555<br>
            传真：+86(10) 58847576<br>
            网址：www.ultrapower.com.cn<br>
            地址：北京市朝阳区北苑路甲13号院北辰泰岳大厦22层  100107<br>
            （地铁5号线北苑路北站，A2出口，向北100米十字路口西南角<br>
      ----------- 居利思义 • 身劳心安 • 人强我强 • 共同发展 -----------<br>
<small>保密提示：本邮件仅供上述收件人使用，并可能包含受法律保护之秘密信息。如果您并非上述所列之收件人，请立即删除本邮件，并勿阅读、传播、复制或以任何方式扩散本邮件。谢谢！</small>
            """
            msg.attach(MIMEText(text, 'html', 'utf-8'))

            for attachment in attachments:
                att = MIMEBase('application', 'octet-stream')
                att.set_payload(open(os.path.join(cfg.TmpPath, attachment), 'rb').read())
                att.add_header('Content-Disposition', 'attachment', filename=Header(attachment, 'utf-8').encode())
                encoders.encode_base64(att)
                msg.attach(att)

            self.server.sendmail(self.Sender, receive + self.Bcc, msg.as_string())
        except Exception as err:
            cfg.log.warning('Mail failed to sent:' + str(err))
            return
        cfg.log.info('[%s %s] sent to %s' % (sub, attachments[0], ','.join(x for x in receive)))


config.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# @File  : config.py
# @Author: Sui Huafeng
# @Date  : 2019/7/25
# @Desc  : 全局变量和公共函数
#

import json
import logging
import os
import sys
import time
from datetime import date, timedelta
from shutil import rmtree

import boto3
import requests


def nextMonth(yy_mm):
    return yy_mm[:2] + '%02d' % (int(yy_mm[2:]) + 1) if yy_mm[2:] < '12' else '%02d01' % (int(yy_mm[:2]) + 1)


# 初始化日志设置
def initLogger(level):
    log = logging.getLogger()
    LogLevel = {'NOTSET': 0, 'DEBUG': 10, 'INFO': 20, 'WARNING': 30, 'ERROR': 40, 'CRITICAL': 50}
    logLevel = LogLevel.get(level.upper(), '30')
    log.setLevel(logLevel)
    return log


# 从S3下载Credit余额，报表模板等配置文件
def downloadMetaData():
    s3 = boto3.resource('s3')
    bucket = s3.Bucket(Bucket)
    for obj in bucket.objects.filter(Prefix='Meta/'):
        if obj.key[-1] == '/':
            continue
        file_name = os.path.join(MetaPath, obj.key.split('/')[-1])
        log.info('Downloading %s' % obj.key)
        bucket.download_file(obj.key, file_name)


Bucket = 'billing-up-972221870813'
BillMonth = os.environ.get('month', (date(date.today().year, date.today().month, 1)
                                     - timedelta(days=10)).strftime('%y%m'))

WorkPath = os.environ.get('path', '/tmp')
MetaPath = os.path.join(WorkPath, 'meta')  # 保存阶梯价格、Credit等配置和消费日志数据
TmpPath = os.path.join(WorkPath, 'tmp')  # 保存中间结果
if os.path.exists(TmpPath):
    rmtree(TmpPath)
    time.sleep(1)  # 防止立刻建立目录出错
[os.mkdir(folder) for folder in (WorkPath, MetaPath, TmpPath) if not os.path.exists(folder)]
log = initLogger(os.environ.get('log', 'WARNING'))

downloadMetaData()

with open(os.path.join(MetaPath, 'Payers.json'), 'r', encoding='utf8') as fp:
    payers = json.load(fp)
with open(os.path.join(MetaPath, 'customers.json'), 'r', encoding='utf8') as fp:
    Customers = json.load(fp)

CustomerNames = {}
for customer_name, customer in Customers.items():
    for awsAccount in customer['Accounts']:
        CustomerNames[awsAccount] = customer_name

"""
libs
"""


# 多级字典中，增加一行的key和默认值
def cascadeDictDefault(dictionary, key_defaults):
    upper = dictionary
    for key, default in key_defaults:
        if key not in dictionary.keys():
            dictionary[key] = default
        upper = dictionary
        dictionary = dictionary[key]
    return dictionary


# 从中行官网爬取当月1日16点左右的美元现汇卖出价
def getUsdExchangeRate(yy_mm):
    try:
        first_work_day = date(year=int('20' + yy_mm[:2]), month=int(yy_mm[2:]), day=1)
        days = first_work_day.weekday() - 4
        if days > 0:
            first_work_day += timedelta(days=3 - days)

        url_parameters = {'erectDate': first_work_day.strftime('%Y-%m-%d 15:00:00'),
                          'nothing': first_work_day.strftime('%Y-%m-%d 16:00:00'),
                          'pjname': 1316, 'page': 1}
        url = 'http://srh.bankofchina.com/search/whpj/search.jsp?%s' \
              % ('&'.join(['%s=%s' % (k, v) for k, v in url_parameters.items()]))
        html = requests.get(url).text.split('\r\n')
        header = ['<select name="pjname" id="pjname">', '<th>货币名称</th>']
        idx = 0
        for lines, line in enumerate(html):  # 找到不连续多行的特征序列
            if header[idx] in line:
                idx += 1
                if idx == len(header):
                    break
        else:
            raise ValueError('no data in html')

        for spacing, line in enumerate(html[lines:]):  # 找出表格所处列数，也就是跟相隔行数
            if '<th>现汇卖出价</th>' in line:
                break
        else:
            raise ValueError('no data in html')

        spacing1 = 0
        lines += spacing
        for line in html[lines:]:  # 按相同列数定位数据行位置
            if '<td>美元</td>' not in line and spacing1 == 0:
                continue
            spacing1 += 1
            if spacing1 > spacing:
                break
        else:
            raise ValueError('no data in html')
    except Exception as err:
        log.warning('Cannot get currency: ' + str(err))
        return 0
    exchange_rate = float(line.split('<td>')[-1].split('</td>')[0]) / 100
    log.info('Get USD exchange rate from BOC website：%f' % exchange_rate)
    return exchange_rate


XlsBill.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# @File  : XlsBill.py
# @Author: Sui Huafeng
# @Date  : 2019/8/28
# @Desc  : 为每个客户生成xls格式的月账单，为收美元的客户生成Invoice
#
import json
import os
from datetime import date
import xlrd
import xlwt
from xlutils import copy
import config as cfg


class XlsBill(object):
    LastDay = {'01': '31', '03': '31', '05': '31', '07': '31', '08': '31', '10': '31', '12': '31',
               '02': '28', '04': '30', '06': '30', '09': '30', '11': '30'}

    def __init__(self, exchange_rate, currency='USD'):
        self.currency = currency
        self.styles = dict(Invoice={}, Bill={})
        self.template = xlrd.open_workbook(os.path.join(cfg.MetaPath, 'Invoice%s.xls' % currency), formatting_info=True)
        self.headers = {self.template.sheet_names()[i]: self.template.sheet_by_index(i).row_values(0)
                        for i in range(1, 2)}
        self.book = copy.copy(self.template)
        self.__setStyle()
        self.exchangeRate = exchange_rate
        self.xlsRow = 1

    def __setStyle(self):
        sheet, styles = self.book.get_sheet('Invoice'), self.styles['Invoice']
        styles.update({col: getStyle(sheet, 11, col) for col in range(1, 4)})
        styles['No'] = getStyle(sheet, 4, 3)
        styles['Date'] = getStyle(sheet, 5, 3)
        styles['Name'] = getStyle(sheet, 6, 2)
        styles['Addr'] = getStyle(sheet, 7, 2)
        styles['Tel'] = getStyle(sheet, 8, 2)
        styles['billMonth'] = getStyle(sheet, 8, 3)
        styles['Amount'] = getStyle(sheet, 26, 3)
        styles['TotalRmb'] = getStyle(sheet, 27, 1)
        styles['AmountRmb'] = getStyle(sheet, 27, 3)

        sheet, styles = self.book.get_sheet('Bill'), self.styles['Bill']
        for layer in range(6):
            styles[layer] = {header: getStyle(sheet, layer, col) for col, header in enumerate(self.headers['Bill'])}

    def run(self, customerName, invoiceNo):
        with open(os.path.join(cfg.TmpPath, '%s.json' % customerName), 'r', encoding='utf-8') as fp:
            monthBills, invoice = json.load(fp)

        self.book = copy.copy(self.template)
        self.fillInvoice(customerName, invoice, invoiceNo)
        self.fillBill(monthBills)
        file_name = os.path.join(cfg.TmpPath, '%s-%s.xls' % (customerName, cfg.BillMonth))
        self.book.save(file_name)  # 保存文件
        cfg.log.info('%s has been saved.' % file_name)

    def fillBill(self, monthBills):
        sheet = self.book.get_sheet('Bill')
        colWidth = [sheet.col(i).width for i in range(10)]
        for account, l1Bill in monthBills.items():
            sheet = self.__addSheet(sheet, colWidth, account)
            self.xlsRow = 0  # 重设xls当前行
            self.__writeLine(sheet, account, l1Bill, 1)
            for product, l2Bill in l1Bill.items():
                self.__writeLine(sheet, product, l2Bill, 2)
                for region, l3Bill in l2Bill.items():
                    self.__writeLine(sheet, region, l3Bill, 3)
                    for l4, l4Bill in l3Bill.items():
                        layer = 5 if 'UsageQuantity' in l4Bill.keys() else 4
                        self.__writeLine(sheet, l4, l4Bill, layer)
                        if 'UsageQuantity' in l4Bill.keys():
                            continue
                        for description, l5Bill in l4Bill.items():
                            self.__writeLine(sheet, description, l5Bill, 5)
            sheet = None

    def __writeLine(self, sheet, key, bill, layer):
        self.xlsRow += 1
        description = 'Account:%s' % key if layer == 1 else key
        description = '  ' * layer + description
        usageQuantity = bill.get('UsageQuantity', '')
        line = (layer, description, usageQuantity, '', bill['TotalCost'])
        bill.pop('TotalCost')
        for header, data in zip(self.headers['Bill'], line):
            writeCell(sheet, self.styles['Bill'][layer][header], data, row=self.xlsRow)

    def __addSheet(self, sheet, colWidth, account):
        if not sheet:
            sheet = self.book.add_sheet(account, cell_overwrite_ok=True)
            for idx, header in enumerate(self.headers['Bill']):
                sheet.col(idx).width = colWidth[idx]
                writeCell(sheet, self.styles['Bill'][0][header], header)
        sheet.panes_frozen = True
        sheet.horz_split_pos = 1

        return sheet

    def fillInvoice(self, customerName, invoice, invoiceNo):
        cust_obj = cfg.Customers[customerName]
        sheet, styles = self.book.get_sheet('Invoice'), self.styles['Invoice']
        sheet.show_headers = False
        sheet.footer_str = b''
        sheet.header_str = b''
        writeCell(sheet, styles['No'], 'P/I 20%s%04d' % (cfg.nextMonth(cfg.BillMonth), invoiceNo))
        writeCell(sheet, styles['Date'], date.today().strftime('%Y-%m-%d'))
        writeCell(sheet, styles['Name'], cust_obj.get('Name', customerName))
        writeCell(sheet, styles['Addr'], cust_obj.get('Addr', ''))
        writeCell(sheet, styles['Tel'], cust_obj.get('Tel', ''))

        first_join_date = cust_obj.get('FirstJoinedTime', '')
        yy, mm = cfg.BillMonth[:2], cfg.BillMonth[2:]
        start_dd = first_join_date[-2:] if first_join_date[:2] == yy and first_join_date[3:5] == mm else '01'
        end_dd = '29' if mm == '02' and int(yy) / 4 == 0 else self.LastDay[mm]
        writeCell(sheet, styles['billMonth'], '20%s/%s/%s - %s/%s' % (yy, mm, start_dd, mm, end_dd))

        for lines, (product, fee) in enumerate(invoice):
            writeCell(sheet, styles[2], product, row=lines + 11)
            writeCell(sheet, styles[3], fee, row=lines + 11)
        writeCell(sheet, styles['Amount'], xlwt.Formula("SUM(D12:D26)"))

        if self.currency == 'RMB2USD':
            writeCell(sheet, styles['TotalRmb'], '应付人民币总额(汇率：%.4f, 税率：17%%)' % self.exchangeRate)
            writeCell(sheet, styles['AmountRmb'], xlwt.Formula("D27*(1+B29)*%.4f" % self.exchangeRate))


def writeCell(sheet, template_cell, value, row=None, col=None):
    row = template_cell['row'] if row is None else row
    col = template_cell['col'] if col is None else col
    sheet.write(row, col, value)
    cell_style = getStyle(sheet, row, col)['style'] if template_cell['style'] else None
    if cell_style:
        cell_style.xf_idx = template_cell['style'].xf_idx


def getStyle(sheet, row, col):
    row_ = sheet._Worksheet__rows.get(row)
    style = row_._Row__cells.get(col) if row_ else None
    return {'row': row, 'col': col, 'style': style}



MonthlyReport.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# @File  : MonthlyReport.py
# @Author: Sui Huafeng
# @Date  : 2019/10/16
# @Desc  : 以代付账户的月度报告MR为数据源，生成客户收据和账单
#  1. 按AWS账户拆分每个payer account的MR
#  2. 修正Credit，只扣减本账号自己的(Payer账号代充共享的扣掉，不管子账号余额)
# 按AWS账户和客户拆分PayerAccount-aws-billing-csv-yyyy-mm.csv
#
import json
import os
import boto3
import csv

if __name__ == '__main__':
    # os.environ['receiver'] = 'sui.hf@163.com'
    # os.environ['month'] = '1910'
    os.environ['log'] = 'DEBUG'

import config as cfg
from XlsBill import XlsBill
from Mail import Mail

Receiver = os.environ.get('receiver', 'maling1@ultrapower.com.cn,jiayf@ultrapower.com.cn').split(',')
SumCols = {'UsageQuantity', 'CostBeforeTax', 'Credits', 'TaxAmount', 'TotalCost'}


class CsvFile(object):
    CsvWriters = {}

    # 按aws账户拆账
    def split(self, payerAccount):
        try:
            fileFullname = downloadAwsMR(payerAccount)
        except Exception as err:
            cfg.log.warning('[%s]downloadAwsMR：%s' % (payerAccount, str(err)))
            return

        accountTotal = getAccountTotal(fileFullname)
        self.__writeItems(fileFullname, accountTotal, payerAccount)

        for linkedAccountId, rowTotal in accountTotal.items():
            customerName = cfg.CustomerNames.get(linkedAccountId, None)
            if not customerName:  # 账号不在配置文件中
                cfg.log.warning('no customer for %s, please update customer.json.' % linkedAccountId)
                continue
            self.__writeSupport(rowTotal, linkedAccountId)
            self.__writeDiscount(rowTotal, linkedAccountId)
            fileName = mrFileName(rowTotal)
            self.CsvWriters[fileName][0].writerow(rowTotal)
        self.__close()

    # 拆分行写入客户csv文件
    def __writeItems(self, file_fullname, accountTotal, payerAccount):
        removeCredit = cfg.payers.get(payerAccount, {}).get('removeCredit', False)
        with open(file_fullname, 'r', encoding='utf-8', newline='') as fp:
            fp.readline()  # 抛弃首行提示信息 Don't see your tags in the report....
            for row in csv.DictReader(fp):
                if row['RecordType'] != 'LinkedLineItem':  # 只拆分资源消费记录
                    continue
                if removeCredit:  # 去掉Credit抵扣，按泰岳折扣算
                    row['Credits'] = '0'
                    row['TotalCost'] = row['CostBeforeTax']
                for col in SumCols:
                    accountTotal[row['LinkedAccountId']][col] += float(row[col])
                fileName = mrFileName(row)
                self.__open(fileName, row) if fileName not in self.CsvWriters else None
                self.CsvWriters[fileName][0].writerow(row)

    # 添加泰岳折扣行，并更新rowTotal(优先使用Customer上的Discount属性，其次使用Account上的)
    def __writeDiscount(self, rowTotal, linkedAccountId):
        cust_obj = cfg.Customers[cfg.CustomerNames[linkedAccountId]]
        linkedAccount = cust_obj['Accounts'][linkedAccountId]
        discountRate = cust_obj.get('Discount', linkedAccount.get('Discount', 0))
        if not discountRate:
            return

        row = rowTotal.copy()
        row['RecordType'] = 'LinkedLineItem'
        row['RecordID'] = ''
        row['ItemDescription'] = '%g%% Discount offered by Ultrapower' % discountRate
        row['ProductName'] = 'Ultrapower Discount'
        row['ProductCode'] = row['UsageType'] = 'UltraDiscount'
        row['UsageType'] = 'NORG-Discount'
        for col in SumCols:
            row[col] = '0'

        discount = - rowTotal['TotalCost'] * discountRate / 100
        rowTotal['TotalCost'] += discount  # 更新汇总行
        row['UsageQuantity'] = str(rowTotal['TotalCost'])
        row['TotalCost'] = str(discount)

        fileName = mrFileName(row)
        self.CsvWriters[fileName][0].writerow(row)
        return

    # 添加aws Support, 并更新rowTotal
    def __writeSupport(self, rowTotal, linkedAccountId):
        cust_obj = cfg.Customers[cfg.CustomerNames[linkedAccountId]]
        linkedAccount = cust_obj['Accounts'][linkedAccountId]
        supportRate = cust_obj.get('SupportRate', linkedAccount.get('SupportRate', 0))
        supportDiscount = cust_obj.get('SupportDiscount', linkedAccount.get('SupportDiscount', 0))
        if not (supportRate or supportDiscount):
            return

        row = rowTotal.copy()
        row['RecordType'] = 'LinkedLineItem'
        row['RecordID'] = ''
        row['ItemDescription'] = 'AWS Support(%g%%)' % supportRate
        row['ProductName'] = 'AWS Premium Support(%g%%)' % supportRate
        row['ProductCode'] = row['UsageType'] = 'OCBPremiumSupport'
        row['UsageType'] = 'NORG-Support'
        for col in SumCols:
            row[col] = '0'

        support = rowTotal['TotalCost'] * supportRate / 100
        rowTotal['TotalCost'] += support  # 更新汇总行
        row['UsageQuantity'] = str(rowTotal['TotalCost'])
        row['TotalCost'] = str(support)

        fileName = mrFileName(row)
        self.CsvWriters[fileName][0].writerow(row)
        return

    # 新建输出文件
    def __open(self, file_name, row):
        fileFullname = os.path.join(cfg.TmpPath, file_name)
        exists = os.path.exists(fileFullname)
        fp = open(fileFullname, 'a', encoding='utf-8', newline='')
        self.CsvWriters[file_name] = (csv.DictWriter(fp, fieldnames=row.keys()), fp)
        if not exists:
            self.CsvWriters[file_name][0].writeheader()

    def __close(self):
        for _, fp in self.CsvWriters.values():
            fp.close()
        self.CsvWriters = {}


class JsonFile(object):
    RegionCode2Name = {'ap-northeast-1': 'Asia Pacific (Tokyo)', 'ap-northeast-2': 'Asia Pacific (Seoul)',
                       'ap-east-1': 'Asia Pacific (Hong Kong)', 'ap-southeast-1': 'Asia Pacific (Singapore)',
                       'ap-southeast-2': 'Asia Pacific (Sydney)', 'ap-south-1': 'Asia Pacific (Mumbai)',
                       'ca-central-1': 'Canada (Central)', 'eu-west-1': 'EU (Ireland)',
                       'eu-central-1': 'EU (Germany)', 'eu-north-1': 'EU (Stockholm)',
                       'eu-west-2': 'EU (London)', 'eu-west-3': 'EU (Paris)',
                       'me-south-1': 'Middle East (Bahrain)', 'sa-east-1': 'South America (Sao Paulo)',
                       'us-gov-east-1': 'AWS GovCloud (US-East)', 'us-gov-west-1': 'AWS GovCloud (US-West)',
                       'us-east-1': 'US East (Northern Virginia)', 'us-east-2': 'US East (Ohio)',
                       'us-west-1': 'US West (Northern California)', 'us-west-2': 'US West (Oregon)',
                       'ap-northeast-3': 'Asia Pacific (Osaka-Local)'
                       }
    RegionAbbr2Name = {'APN1': 'Asia Pacific (Tokyo)', 'APN2': 'Asia Pacific (Seoul)',
                       'APE1': 'Asia Pacific (Hong Kong)', 'APS1': 'Asia Pacific (Singapore)',
                       'APS2': 'Asia Pacific (Sydney)', 'APS3': 'Asia Pacific (Mumbai)',
                       'CAN1': 'Canada (Central)', 'EU': 'EU (Ireland)',
                       'EUC1': 'EU (Germany)', 'EUN1': 'EU (Stockholm)',
                       'EUW2': 'EU (London)', 'EUW3': 'EU (Paris)',
                       'MES1': 'Middle East (Bahrain)', 'SAE1': 'South America (Sao Paulo)',
                       'UGW1': 'AWS GovCloud (US-West)', 'USE1': 'US East (Northern Virginia)',
                       'USE2': 'US East (Ohio)', 'USW1': 'US West (Northern California)',
                       'USW2': 'US West (Oregon)', 'APN3': 'Asia Pacific (Osaka-Local)',
                       'UGE1': 'AWS GovCloud (US-East)'
                       }
    CloudFrontRegionAbbr2Name = {'AP': 'Asia Pacific', 'AU': 'Australia', 'CA': 'Canada', 'EU': 'Europe',
                                 'IN': 'India', 'JP': 'Japan', 'ME': 'Middle East', 'SA': 'South Africa',
                                 'ZA': 'South America', 'US': 'United States'
                                 }

    # 汇聚成客户的总账单
    def merge(self, customerName):
        consolidateRows = {}
        accounts = set()  # 实际产生费用的账户
        fileName = mrFileName('', customerName)
        fileFullname = os.path.join(cfg.TmpPath, fileName)
        row = None
        with open(fileFullname, 'r', encoding='utf-8', newline='') as fp:
            for row in csv.DictReader(fp):
                if row['RecordType'] != 'LinkedLineItem':
                    continue
                accounts.add(row['LinkedAccountId'])
                key = (row['ProductCode'], row['UsageType'], row['ItemDescription'])
                for col in SumCols:
                    row[col] = float(row[col])

                if key not in consolidateRows:
                    consolidateRows[key] = row
                else:
                    for col in SumCols:
                        consolidateRows[key][col] += row[col]

        if len(accounts) < 2:
            return  # 只有一个有费用账号
        with open(fileFullname, 'a', encoding='utf-8', newline='') as fp:
            csvWriter = csv.DictWriter(fp, fieldnames=row.keys())
            statementTotalRow = getStatementTotal(row.copy())
            for consolidateRow in consolidateRows.values():
                consolidateRow['RecordType'] = 'PayerLineItem'
                consolidateRow['LinkedAccountId'] = consolidateRow['LinkedAccountName'] = ' All'
                for col in SumCols:
                    statementTotalRow[col] += consolidateRow[col]
                    consolidateRow[col] = str(consolidateRow[col])
                csvWriter.writerow(consolidateRow)

            for col in SumCols:
                statementTotalRow[col] = str(statementTotalRow[col])
            csvWriter.writerow(statementTotalRow)

    # 汇总成靠近AWS系统内形式的分层的的月账单
    def invoice(self, customerName):
        monthBills = {}
        fileName = mrFileName('', customerName)
        fileFullname = os.path.join(cfg.TmpPath, fileName)
        with open(fileFullname, 'r', encoding='utf-8', newline='') as fp:
            [self.add(monthBills, row) for row in csv.DictReader(fp) if row['RecordType'][-8:] == 'LineItem']

        monthBill = monthBills[' All'] if ' All' in monthBills else list(monthBills.values())[0]
        invoice = [(product, round(v['TotalCost'], 2))
                   for product, v in monthBill.items() if product != 'TotalCost' and abs(v['TotalCost']) > 0.005]
        invoice.sort(key=lambda v: abs(v[1]), reverse=True)
        if len(invoice) > 15:  # 最多保留15行
            others = sum(v for d, v in invoice[14:])
            invoice[14] = ('Others', 0)
            invoice = invoice[:15]
        else:
            others = None
        invoice.sort(key=lambda v: v[1], reverse=True)
        if others:  # 保证其他在所有正值行之后
            invoice[invoice.index(('Others', 0))] = ('Others', others)

        with open(os.path.join(cfg.TmpPath, '%s.json' % customerName), 'w', encoding='utf-8') as fp:
            json.dump((monthBills, invoice), fp, indent=2, sort_keys=True)

    def add(self, bill, row):
        productName = row['ProductName'].split('AWS ', 1)[-1].split('Amazon ')[-1]
        regionName = self.__getRegion(row)
        l4Name = self.__l4Label(row)
        for layer in (row['LinkedAccountId'], productName, regionName, l4Name, row['ItemDescription']):
            if not layer:  # 不需要l4Name这一级
                continue
            cfg.cascadeDictDefault(bill, [(layer, {}), ('TotalCost', 0)])
            bill = bill[layer]
            bill['TotalCost'] += float(row['TotalCost'])
        cfg.cascadeDictDefault(bill, [('UsageQuantity', 0)])
        bill['UsageQuantity'] += float(row['UsageQuantity'])

    # 从账单行中尽量抽取region信息，向AWS系统中账单层次靠拢
    def __getRegion(self, row):
        if float(row['TotalCost']) < 0:
            return ' No Region'
        availabilityZone = row.get('AvailabilityZone', '')
        if availabilityZone:
            regionCode = availabilityZone if availabilityZone[-1] <= '9' else availabilityZone[:-1]
            regionName = self.RegionCode2Name.get(regionCode, '')
            if regionName:
                return regionName

        productCode = row['ProductCode']
        splitedUsageType = row['UsageType'].split('-')
        if productCode == 'AWSGlobalAccelerator':  # AP/NA/Global
            regionName = splitedUsageType[0]
        elif productCode == 'awskms':  # usageType中带着regionCode
            regionCode = 'Global' if len(splitedUsageType) < 3 else '-'.join(splitedUsageType[:3])
            regionName = self.RegionCode2Name.get(regionCode, 'US East (Northern Virginia)')
        elif productCode == 'AmazonCloudFront':
            regionName = self.CloudFrontRegionAbbr2Name.get(splitedUsageType[0], 'Global')
        else:
            regionName = self.RegionAbbr2Name.get(splitedUsageType[0], 'US East (Northern Virginia)')
        return regionName

    RdsL4 = {
        'backup storage': ' Backup Storage', 'rovisioned ': ' Provisioned Storage',
        ' storage ': ' Storage and I/O', ' I/O requests ': ' Storage and I/O',
        'Aurora': ' for Aurora', 'MySQL': ' for MySQL Community Edition'
    }

    def __l4Label(self, row):
        product_name = row['ProductName']
        operation = row['Operation']
        usage_type = row['UsageType']
        line_item_description = row['ItemDescription']

        if product_name == 'Amazon Relational Database Service':
            for k, v in self.RdsL4.items():
                if k in line_item_description:
                    type_ = product_name + v
                    break
            else:
                type_ = product_name

            if 'hour (' not in line_item_description and 'for ' in type_:
                type_ += ' Reserved Instances'

        elif product_name == 'Amazon Elastic Compute Cloud':
            type_ = operation if 'LoadBalancing' in operation or 'NatGateway' in operation \
                else 'EBS' if 'EBS' in usage_type \
                else 'Elastic IP Addresses' if 'ElasticIP' in usage_type \
                else 'Amazon Elastic Compute Cloud '
            if type_ == 'Amazon Elastic Compute Cloud ':
                type_ += 'Running Linux/UNIX' if 'Linux' in line_item_description \
                    else 'Running Windows' if 'Windows' in line_item_description \
                    else ''
                type_ += ' Spot Instances' if 'SpotUsage' in usage_type \
                    else ' Reserved Instances' if 'On Demand' not in line_item_description \
                    else ''
        elif product_name in ('Amazon ElastiCache', 'Amazon Redshift'):
            type_ = product_name + ' ' + operation
            if ' instance' in line_item_description:
                type_ += ' Reserved Instances'
        elif product_name == 'Amazon Elastic File System' and operation == 'Storage':
            type_ = product_name + ' Standard Storage Class'
        elif product_name == 'AWS Shield':
            type_ = product_name + ' ' + usage_type if usage_type == 'Shield-Monthly-Fee' else 'Bandwidth'
        else:
            type_ = None

        if float(row['TotalCost']) < 0:
            type_ = None

        return type_


# 下载AWS MR文件
def downloadAwsMR(payerAccount):
    bucket = boto3.resource('s3').Bucket(cfg.Bucket)
    fileName = '%s-aws-cost-allocation-20%s-%s.csv' % (payerAccount, cfg.BillMonth[:2], cfg.BillMonth[2:])
    fileFullname = os.path.join(cfg.TmpPath, fileName)
    cfg.log.info('Downloading [%s]%s(MR)' % (cfg.Bucket, fileName))
    bucket.download_file(fileName, fileFullname)
    return fileFullname


def mrFileName(row, customerName=None):
    customerName = customerName if customerName else cfg.CustomerNames.get(row['LinkedAccountId'], 'Missing')
    fileName = '%s-%s.csv' % (customerName, cfg.BillMonth)
    return fileName


# 重新计算各账户的汇总数据
def getAccountTotal(file_fullname):
    accountTotal = {}
    with open(file_fullname, 'r', encoding='utf-8', newline='') as fp:
        fp.readline()  # 抛弃首行提示信息 Don't see your tags in the report....
        for row in csv.DictReader(fp):
            if row['RecordType'] != 'AccountTotal':
                continue
            for col in SumCols:
                row[col] = 0
            accountTotal[row['LinkedAccountId']] = row
    return accountTotal


def getStatementTotal(row):
    row['RecordType'] = row['RecordID'] = 'StatementTotal'
    row['LinkedAccountId'] = row['LinkedAccountName'] = ''
    row['InvoiceID'] = row['InvoiceDate'] = ''
    row['ProductCode'] = row['ProductName'] = row['UsageType'] = ''
    row['Operation'] = row['RateId'] = ''
    row['UsageStartDate'] = row['UsageEndDate'] = row['UsageQuantity'] = ''
    period = '%s - %s' % (row['BillingPeriodStartDate'], row['BillingPeriodEndDate'])
    row['ItemDescription'] = 'Total statement amount for period ' + period
    for col in SumCols:
        row[col] = 0
    return row


def sendBill():
    with Mail() as mail:
        for customerName, cust_obj in cfg.Customers.items():
            file_names = ['%s-%s.xls' % (customerName, cfg.BillMonth)]
            send_csv = cust_obj.get('Csv', 0)
            if send_csv:
                file_names.append(mrFileName('', customerName))
            subject = 'AWS账单(20%s-%s)%s' % (cfg.BillMonth[:2], cfg.BillMonth[2:], customerName)
            text = '尊敬的 %s 客户:<br>附件是贵司(20%s-%s)月AWS账单。请查收<br>' \
                   % (customerName, cfg.BillMonth[:2], cfg.BillMonth[2:])
            mail.send(Receiver, file_names, sub=subject, text=text)


def run():
    csvFile = CsvFile()
    for payerAccount in cfg.payers:
        csvFile.split(payerAccount)

    exchangeRate = cfg.getUsdExchangeRate(cfg.nextMonth(cfg.BillMonth))
    invoieNo = 0
    jsonFile = JsonFile()
    for customerName, cust_obj in cfg.Customers.items():
        fileName = mrFileName('', customerName)
        fileFullname = os.path.join(cfg.TmpPath, fileName)
        if not os.path.exists(fileFullname):
            cfg.log.warning('%s not exists, please confirm.' % fileFullname)
            continue
        if len(cust_obj['Accounts']) > 1:  # 多账号才汇总
            jsonFile.merge(customerName)
        jsonFile.invoice(customerName)  # 生成分层账单数据和收据数据
        currency = cust_obj.get('Currency', 'USD').upper()
        invoieNo += 1
        XlsBill(exchangeRate, currency).run(customerName, invoieNo)  # 生成xls账单
    sendBill()  # 邮件发送账单


# AWSLamda调用的入口函数
def lambda_handler(event, context):
    run()
    return {
        'statusCode': 200,
        'body': json.dumps('UltraInvoice Finished!')
    }


# EC2上运行的入口
if __name__ == '__main__':
    run()
